fix(logging): redact JWT-like usernames in logs

IceHamster opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->
- Add regex-based detection for JWT-like strings (header.payload.signature format).
- When sanitizing objects, redact the username field if it contains a JWT.
- Add a unit test to verify redaction works as expected.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

Some registries (e.g., [NGINX Plus](https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-docker/#set-up-docker-for-the-f5-container-registry)) require using a JWT as the username when logging in:
```sh
docker login private-registry.nginx.com \
  --username=<JWT_FROM_MYF5> \
  --password=none
```
Our Renovate config mirrors this via hostRules:
```js
{
  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "private-registry.nginx.com",
      "username": "<output_of_jwt_token>",
      "password": "none"
    }
  ]
}
```
Because Renovate normally assumes username is a non-secret identifier, it was not sanitized by default.
This leads to JWT credentials being logged in plain text. This PR adds special handling to redact JWT-like values when they appear in username.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

IceHamster committed:
chore: Mask username when an JWT token

viceice reviewed (commented):
please describe why the username field isn't already sanitized. how do you configure the token?

IceHamster commented:
@viceice I have updated the _Context_ in the PR Description.

viceice reviewed (changes requested):
then this should be done somewhere here

https://github.com/renovatebot/renovate/blob/7cf05ac34dd63a970e0af11327d2fc44b2609e1d/lib/util/host-rules.ts#L66

so the username is added to sanitizer list

looperunbalanced commented on lib/logger/utils.ts:
```diff
@@ -216,6 +216,9 @@ export function sanitizeValue(
   }
 
+  // Define a regular expression to detect the JWT format (header.payload.signature)
+  const jwtRegex = /^[\w-]+\.[\w-]+\.[\w-]+$/;
```
This pattern would yield a significant amount of false positives, doesn't it?

RFC-7515 Appendix C does seem to indicate that and underscore is also a valid character in the token.

IceHamster replied:
Good point, thank you. The previous logic could cause false positives.

I've pushed a commit that tightens the validation. The new logic uses a more specific regex and adds a minimum length check, since JWTs are almost always significantly longer than usernames.

IceHamster commented:
> then this should be done somewhere here
> 
> https://github.com/renovatebot/renovate/blob/7cf05ac34dd63a970e0af11327d2fc44b2609e1d/lib/util/host-rules.ts#L66
> 
> so the username is added to sanitizer list

Thanks for the suggestion! That's a good idea, but i think there's a timing issue here.

The raw configuration object gets logged at the debug level before the host-rules.ts logic has a chance to parse the rules and add the username to the sanitizer. So the JWT is already in the logs by the time that code runs.

Here's a snippet from the logs that shows this:

<details>
<summary>Click to expand debug log</summary>
<code>
DEBUG: Using RE2 regex engine
DEBUG: Parsing configs
# ...
DEBUG: File config
       "config": {
         "hostRules": [
           {
             "hostType": "docker",
             "matchHost": "private-registry.nginx.com",
             "username": "ey......",
             "password": "***********"
           },
         ]
       }
# ...
DEBUG: Found valid git version: 2.51.0
DEBUG: Setting global hostRules
DEBUG: Adding password authentication for private-registry.nginx.com (hostType=docker) to hostRules
</code>
</details>

IceHamster committed:
fix(logger): Tighten JWT detection in username field

Updated the regex to be more specific to Base64URL and added a length
check to prevent redacting valid usernames. Added a test for the
false-positive case.

looperunbalanced commented on lib/logger/utils.ts:
```diff
@@ -216,6 +216,9 @@ export function sanitizeValue(
   }
 
+  // Define a regular expression to detect the JWT format (header.payload.signature)
+  const jwtRegex = /^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$/;
```
Are you deliberately making the signature optional?

looperunbalanced commented on lib/logger/utils.ts:
```diff
@@ -230,6 +233,13 @@ export function sanitizeValue(
         }
+      } else if (
+        key === 'username' &&
+        val.length > 60 &&
+        is.string(val) &&
```
Shouldn't these conditions be in the opposite order?

IceHamster replied:
Yeah, you are probably right that it would be better to check that the value is an string, before checking
the length. Will push an fix.

IceHamster committed:
fix(logger): reordering conditions for

viceice reviewed (commented):
please also add it on host rules add function

RahulGautamSingh reviewed (approved):
LGTM except host rules comment

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction

IceHamster committed:
fix(logger): Sanitize username if JWT in hostRules

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction

astellingwerf commented on lib/util/host-rules.ts:
```diff
@@ -38,6 +38,8 @@ export function migrateRule(rule: LegacyHostRule & HostRule): HostRule {
   return result;
 }
 
+const jwtRegex = /^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$/;
```
I imagine we ought to use re2 here instead.

IceHamster committed:
fix(logger): Use re2 for regex)

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction

viceice commented on lib/util/host-rules.ts:
```diff
@@ -63,6 +66,16 @@ export function add(params: HostRule): void {
+  if (
+    is.string(rule.username) &&
+    rule.username.length > 60 &&
+    jwtRegex.test(rule.username)
```
I'm thinking if should always redact usernames,  just in case there are other formats

IceHamster replied:
@viceice We could then add `username` to `confidentialFields, so we do not need extra logic to mask the username.

Would it create other issues if we mask the username?

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction

IceHamster committed:
feat(logger): Treat username as a confidential field

IceHamster committed:
refactor(host-rules): remove unused import

IceHamster commented on lib/logger/utils.ts:
They are logged before they are set to "confidential"
But we can mask any username, and not only JWT tokens

RahulGautamSingh replied:
Yes, let be consistent

viceice replied:
add to `redactFields` list instead

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction

IceHamster committed:
chore(npmrc.spec.ts): Updated tests

RahulGautamSingh commented on lib/logger/utils.ts:
```diff
@@ -216,6 +216,9 @@ export function sanitizeValue(
   }
 
+  // Define a regular expression to detect the JWT format (header.payload.signature)
+  const jwtRegex = /^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$/;
```
Weird, it showed resolved on my screen 🤨

viceice reviewed (commented):
see open discussion

IceHamster committed:
Merge branch 'main' into fix/jwt-username-redaction
